<a href="https://colab.research.google.com/github/NumericalAnalysis-YTU/HEP/blob/main/cpp1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

The C and C++ programming languages are quite different from each other, even though they share some common syntax.

 First C++ Program

Pedagogical and traditional programming books begin by writing the words Hello World to the screen. 

In [2]:
!g++ --version


g++ (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Copyright (C) 2017 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.



In [11]:
%%writefile hello.cpp
/* First C++ program*/
#include <iostream>
using namespace std;

int main()
{
  cout << "Hello World!";
  return 0;
}

Overwriting hello.cpp


In [15]:
!g++ hello.cpp -o hello_run; ./hello_run

Hello World!

In [21]:
%%writefile temp.cpp
#include <iostream>
int main() {
  std::cout << "hello world \n \t First program of C++" << "\n";
  std::cout << "on FIZ5122";
  return 0;
}

Overwriting temp.cpp


In [23]:
!g++ temp.cpp -o temp_run; ./temp_run

hello world 
 	 First program of C++
on FIZ5122

What does it do?	

Character

Newline	\n
Vertical tab	\v
Backslash	\\
Carriage return	\r
Horizontal tab	\t
Backspace	\b
Question mark	\?
Formfeed	\f
Alert (bell)	\a
Double quote	\”
Single quote	\’

VARIABLES

An identifier is a named memory location.
A variable is an identifer that C++ programs use to store values needed by your program.
Variable names in C++ can consists of letters, numeric digits, the the underscore character, but they can not begin with a digit. When naming variables, remember that you can not you a C++ keyword. A keyword is a word that the compiler reserves with a special meaning. The length of variable names are limited to 255 characters.

DECLARING A VARIABLE

You must declare a variable before you can use them

Syntax:    Data type   variable name(s);

Example:      int x,y ;

You can initialize a C++ variable when you declare it.

Example:     double  x=4.5;

The first column shows the source code line numbers, these do not actually appear in the source code. The C++ source code for our first program is shown in the second column. The last column shows the output of the program, that is the result of the program once compiled and executed.

This very simple C++ program is saved in a file called 01hello.cpp. The program outputs the expression Hello World! to the user's screen. Note that the name of a C++ program file can have the extension .cpp, .c++, .cxx or .cc.

Some guides to compiling and running C++ programs: Compiling C++ programs and The Dev-C++ compiler.

If you've installed a C++ compiler and set up the proper development environment, you should be able to compile this C++ program and make it into an executable file. A compiler translates the statements of program written in a *high level language* into a *low level language*, the machine code. Each computer language has its own compiler. Examples are: Fortran, C, C++, Java, Pascal, Basic.

Now, we will look at the code line by line.

* Line 1: // First C++ program

This is a comment line. All lines starting with two slash signs (//) are considered as comments. That is such lines are ignored by the compiler, and they do not have any effect on the executable. Comments in programs help the programmer (and users) to understand what the program (or section) does. This becomes especailly important for large program sources, or sources that are part of a larger project where more than one programmer is developing or using the code.

* Line 2: #include <iostream>

Lines beginning with a hash (or pound) sign (#) are directives for the preprocessor. The preprocessor runs before the compiler each time the compiler is invoked. The preprocessor translates any line that begins with a hash symbol into a special command, getting your code file ready for the compiler. Here, #include <iostream> tells the preprocessor to include the iostream standard file which contains the declarations of the basic standard input-output library in C++.

This specific file (iostream) can be found usually under the folder:
for Linux (GCC): /usr/include/
for Windows (Dev-C++): C:\Dev-Cpp\include\

* Line 3: using namespace std;

A namespace allows to group entities like variables, classes, objects and functions under a name. All elements belonging to the standard C++ library are declared in what is called a std namespace. For example, cout is defined under std namespace where you can see the details in the file <iostream>. This line will be included in most of the source codes included in these notes. If you omit this line, in order to use cout, you need to write the name of the namespace (std) followed by scope operator (::) before cout. In fact, you have to put std:: before using any elements of std namespace. Such as, std::cout << "Hello world!";.

* Line 4:

An empty line does nothing except for help the programmer to view the source code more clearly.

* Line 5: int main()

The actual program starts with a function named main(). Every C++ program has a main() function. In general, a function is a block of code that performs one or more actions. Usually functions are invoked or called by other functions, but main() is special. When your program starts, first, main() is called automatically.

In C++, each function has a data type. Here data type of the main() is selected as int (integer). That is, at the end of execution of the main program, main() will return an integer number. Functions and returning values are discussed in detail on Section 5.

*Lines 6 and 9:

The body of the main() function is enclosed in braces ({}).

Line 7: cout << "Hello World!";
This line is a C++ statement which performs the only action that generates a visible effect in our first program. Each statement must end with a semicolon character (;). Here's how cout is used: type the word cout, followed by the output insertion (or redirection) operator (<<). Whatever follows the output insertion operator is output to the screen. If you want to output a string of characters be sure to enclose them in double quotes (" "), as shown on line 7, "Hello World!".

* Line 8: return 0;

The return statement causes the main() function (i.e. the program) to finish. All ANSI-compliant programs declare main() to return an int. This value is "returned" to the operating system when your program completes.

Some programmers signal an error by returning the value different from 0 (in general 1 or -1). However, in these notes, main() will in general return 0, i.e. we will not generate error signals.

Finally note that lines 7 and 8 are indented by two spaces; this is for clarity and you can use any number of spaces you like. This becomes more important as your program grows in size.

Comments

C++ supports two types of comments:
// line comment
The C++ compiler ignores everything after and on the same line of a double-slash.

/* block comment */

C++ ignores everything between the opening comment mark (/*) and closing comment mark (*/). This block of comments may span more than one line. This is actually taken from C Programming Language.

Example usages are given below.

  // This comment does not increase the size of

  // the executable file (binary code), nor does

  // it affect the performance speed.
  /*

     This comment does not increase the size of the executable file (binary code), nor does it affect the performance speed.
  */

C++ Variables and Types: Int, Char, Float, Double, String & Bool



* int
* double
* char
* "string"
* bool (true, false) 


Variables in C++

A C++ variable provides us with a named storage capability. It allows the programmer to manipulate data as per the need. Every variable has a type in C++. 

The variable type helps to determine the size and layout of the variable’s memory map, the range of values that can be stored within that memory, and the set of operations that can be applied to it.

* Int:

An integer is a numeric literal (associated with numbers) without any fractional or exponential part. Example. 120, -90, etc.

* Double:

It is a double-precision floating point value. Example: 11.22, 2.345

* Char:

A character literal is created by enclosing a single character inside single quotation marks. For example: ‘a’, ‘m’, ‘F’, ‘P’, ‘}’ etc.

* Float:

A floating-point literal is a numeric literal that has either a fractional form or an exponent form. For example: 1.3, 2.6

* String Literals:

A string literal is a sequence of characters enclosed in double-quote marks. For example: “How are you?”

* Bool:

It holds Boolean value true or false.


Rules of Declaring Variables in C++

Here are some common rules for naming a variable:

A C++ variable name can only have alphabets, numbers, and underscore.

A C++ variable name cannot begin with a number.

Variable names should not begin with an uppercase character.

A variable name used in C++ cannot be a keyword. For example, int is a keyword that is used to denote integers.

A C++ variable name can start with an underscore. However, it is not considered a good practice.

Identifiers must

begin with either a letter or an underscore (‘_’).

And are case-sensitive; upper and lowercase letters are distinct:

// defines four different int variables

int guru99, gurU99, GuRu99, GURU99;

Basic Input and Output (I/O)

The standard C++ library includes the header file iostream, where the standard input and output stream objects are declared. One can see 'streams in C++' how to use:

* cout to output data to the screen and
* cin to input data from the keyboard.

Streams are out of scope of this basic tutorial.

Standard Output

By default, the standard output of a program is the screen, the C++ stream object defined to access it is cout. For now, you can use cout without fully understanding how it works. To output a value to the screen, we need to use cout, followed by the insertion operator (<<), which you create by typing the less-than character (<) twice. Even though this is two characters, C++ treats it as one.

Here are some basic usage examples:

  cout << "Hello World!";    // Outputs 

"Hello World!" on the screen

  cout << 1299;              // Outputs 

the number 1299 on the screen

  cout << x;                 // Outputs 

the content of x on the screen
The insertion operator (<<) may be used more than once in a single statement:

  cout << "Hi, " << "this is " << "a 

simple C++ statement";
Note that, cout does not add a line break. For example:

  cout << "Computer Programming ";
  cout << "is important for engineers.";

will be output as

  Computer Programming is important for engineers.

A line break can be added by inserting a '\n' character or a using a endl manipulator.

  cout << "Computer Programming \n";
  cout << "is important for engineers.";

or

  cout << "Computer Programming " << endl;

  cout << "is important for engineers.";

These two produce the following (same) output:

*  Computer Programming is 
*  important for engineers.

Standard Input

By default, the standard input device is the keyboard. The C++ stream object defined to access it is cin. Handling the standard input is done by applying the overloaded operator of extraction (>>) on the cin stream.

A value can be input from the keyboard and assigned to a variable x as follows:

  cin >> x;

and two variables can assigned as follows:
 *  cin >> x;
 *  cin >> y;
or
  cin >> x >> y;
The following program summaries the use of cout and cin; two integer values are input and their sum is output.

In [ ]:
%%writefile temp.cpp
#include <iostream>

int main() {
  for (int i = 0; i < 10; i++) std::cout << "Hello World" << "\n";
}

In [ ]:
!g++ temp.cpp; ./a.out

In [ ]:
%%writefile temp.cpp
// Calculates the sum of two integers
#include <iostream>
using namespace std;

int main()
{
  int a, b, total;
  cout << "Enter two integers to be added: ";
  cin >> a >> b;
  total = a + b;
  cout << "The sum is " << total << endl;

  return 0;
}

In [ ]:
!g++ temp.cpp; ./a.out

In [27]:
%%writefile temp.cpp
#include <iostream>

int main() {
    std::cout << "Hello World! \n\n";
    std::cout << "I am learning C++";
    return 0;
}

Overwriting temp.cpp


In [28]:
!g++ temp.cpp; ./a.out

Hello World! 

I am learning C++

Three integer variables are declared on line 7 (we will see the details of the variables in the next section). At line 9 variables a and b are assigned with values that are input from the keyboard, their sum is assigned to the variable total on line 10, and the contents of total is output on line 11.


Examples

The following program is another example of basic input and output.

Note the use of std:: (removing the need for the using namespace statement), and the omission of the return statement (which is optional).

In [24]:
%%writefile temp.cpp
// Calculates the area of a triangle
// given an input of the triangle base and height.
#include <iostream>

int main()
{
  float base, height, area;
  std::cout << "Enter the base and height of the triangle: ";
  std::cin >> base >> height;
  area = 0.5*base*height;
  std::cout << "The area of the triangle is " << area << std::endl;
}

Overwriting temp.cpp


In [26]:
!g++ temp.cpp; ./a.out

Enter the base and height of the triangle: 2 3
The area of the triangle is 3


|Operator Name	|Operator Symbol|	Example	| Comment
|--- | ---| ---|---
|Addition |	+	 |num1 + num2	|.
|Subtraction	|-|	num1 - num2	|.
|Multiplicaton	|*  |	num1*num2	|.
|Division |	/	|15/3 or  15.0/2.0 |Integer division; result is 7; fraction is truncated  / Floating-point division; result is 7.500000 |
|Modulus|	%	|5%2|	Performs the division and finds the remainder; the result is 1
|Unary minus	-	-(num1)	if value of num1 is 10, then -(num1) is -10
|Assignment	|=|	num1=10;	|assigns 10 to variable num1|
|+=|	num += 10;	|Equivalent to num = num +10;|
|-=|	num -= 10;  |	Equivalent to num = num -10;|
|*=	|num |*= 10;	|Equivalent to num = num *10;|
|/=	|num /= 10;	  |Equivalent to num = num /10;|
|%=	|num %= 10;	  |Equivalent to num = num %10;|
|Increment|	++	  |num++ or ++num|	Increment num to 1|
|Decrement	|--   |num-- or --num|	Decrement num to 1|